### ISSUE TITLE
proposal: solving major problem with multi-language node injections/extensions

### ISSUE THREAD

**Issue by AlbertMarashi** [tree-sitter/tree-sitter#3625]
### Problem

Currently, language injections are treated as the top-level node in that language's tree-sitter.

It would be very valuable to provide a feature that allowed for treating a specific injection as a specific AST node in that language.

## Example 1: Svelte

**Generic definitions**

In Svelte, we define component generics in the following way:

```svelte
<script generics="T extends string">
```

Unfortunately, if we treat the attribute value string as injected typescript code, we get an error in the typescript parser:

![image](https://github.com/user-attachments/assets/1ce169d4-6cea-4d65-8d12-54610e67c6f0)

This is because `T extends string` is not a valid top-level typescript statement.

If we could instead tell tree-sitter to interpret this as as a [`type_parameter`](https://github.com/tree-sitter/tree-sitter-typescript/blob/master/common/define-grammar.js#L1005) node within the tree-sitter-typescript, it would solve our issues.

**Snippet statements**
In the following example:
```svelte
{#snippet foo({ bar, baz })
```
Treating the code `foo({ bar, baz })` as typescript code would result in a function call expression.

In our case, it would be more appropriate to treat it as a function definition expression (without the `function ` prefix)

We could achieve this by targeting some of the nodes within [`function_declaration`](https://github.com/tree-sitter/tree-sitter-javascript/blob/master/grammar.js#L753C5-L753C25) inside of the javascript grammar.js

**Each blocks**
```svelte
{#each foo.thing as { bar, baz }, index (key)}
```

A svelte each block's ast could be parsed in the following way (simplified)
```ts
import javascript_grammar from "..."

module.exports = grammar({
  each_start: $ => seq(
    "{",
    "#",
    keyword("each"),
    field("iterator", sublanguage(javascript_grammar, "expression")),
    keyword("as"),
    choice(
      field("pattern", sublanguage(javascript_grammar, "destructure_pattern")),
      field("identifier", sublanguage(javascript_grammar, "identifier")),
    )
    $.each_key
  ),
  each_key: $ => optional(seq(
    "(",
    field("identifier", sublanguage(javascript_grammar, "identifier")),
    ")"
  ))
})

function keyword(text: string) {
   ...
}
```

The `sublanguage` feature could tell tree-sitter to automatically know how to treat a specific AST, and to treat it as an injection of a specific subnode.

This would naturally solve huge amounts of issues, and make inter-language extensibility a breeze to work with

### Example 2: CSS styles

Assume we have the following CSS code and we want to treat the `style` attribute as sublanguage targetting the CSS `declaration`
```html
<div style="background: red; color: blue">
```

```ts
import css_grammar from "..."

// HTML grammar, or something
module.exports = grammar({
  css_declarations_list: $ => seq(sublanguage(css_grammar, "declaration"),
  style_value $ => seq(
     "\"",
     $.css_declarations_list,
     "\"",
  //...
})
```

### Example 3: Markdown language node support

Often, as developers, when quoting markdown, we quiet often want to support the ability to target a specific node of a language

Lets assume we have the following markdown code

````md
Some json:
```json
"foo": "bar"
```
````

Unfortunately, this is invalid JSON. Wouldn't it be nice if we could do something like:

````md
Some json:
```json->property
"foo": "bar"
```
````

And have language parsers understand to treat this as a `property` node within the JSON syntax highlighter?

### Example 4: TSX extends typescript extends javascript

Currently, TSX support for https://github.com/tree-sitter/tree-sitter-typescript is supported via common shared grammar, with the grammars being conditionally adjusted based on the dialect passed in.

What if instead, the grammar for `TSX` support was much simpler, and just referenced the typescript language, overrode some types, and injected it's own nodes like into the grammar selectively performing minor grammar surgery on the grammars to add support for TSX nodes.

```ts
      jsx_opening_element: $ => prec.dynamic(-1, seq(
        $._jsx_start_opening_element,
        '>',
      )),
```

Additionally, the same sort of behavior could be added to have typescript be a direct extension of javascript.

This is sort of done with the grammar extension as shown here, which works okay, but it's not as powerful as I think it could be.

```ts
module.exports = function defineGrammar(dialect) {
  return grammar(JavaScript, {
    name: dialect,
```

### Example 5: SurrealDB embedded javascript functions
SurrealDB allows for advanced functions with complicated logic, by allowing embedded functions to be written in JavaScript.

```surql
CREATE person SET scores = function() {
  return [1,2,3].map(v => v * 10);
};
```

In this case, we might specifically want to allow only for parsing of javascript functions, instead of any arbitary javascript node.

Currently, one would need to implement scanner.c logic which performs bracket matching or copying/rewriting the javascript grammar to implement the nodes that were required, which can be error prone and expensive for authors to maintain.

---

The above summarizes my frustrations and suggestions for what is missing from tree-sitter.

I believe adding the described features would make tree-sitter a truly universal and powerful programming tool.

## Related issues
- [ ] https://github.com/tree-sitter/tree-sitter/discussions/793
- [ ] https://github.com/tree-sitter/tree-sitter/issues/121
- [ ] https://github.com/tree-sitter-grammars/tree-sitter-svelte/issues/10
- [ ] https://github.com/helix-editor/helix/discussions/11157
- [ ] https://github.com/nvim-treesitter/nvim-treesitter/issues/1111
- [ ] https://github.com/nvim-treesitter/nvim-treesitter/issues/2212
- [ ] + many more

**Comment by clason** [comment]
Can't you write a "single-statement" grammar that you can inject into HTML _and_ into "proper" svelte? Or simply change the grammar so that single statements are valid at the top level? (You're writing a parsing tool to use in an editor, not a formal language syntax for verification.)

**Comment by AlbertMarashi** [comment]
> Can't you write a "single-statement" grammar that you can inject into HTML _and_ into "proper" svelte? Or simply change the grammar so that single statements are valid at the top level? (You're writing a parsing tool to use in an editor, not a formal language syntax for verification.)

I don't exactly follow.

The svelte generics example (`T extends string`) would not be a valid typescript top-level statement (we are deferring to the typescript language with `@language` and `@content` inside of the svelte extension's `injections.scm`

We just have no way of telling the typescript treesitter that we're expecting a type parameter here, instead of a top level typescript declaration

**Comment by sogaiu** [comment]
FWIW:

> Can't you write a "single-statement" grammar that you can inject into HTML and into "proper" svelte?

I think this kind of thing is something the Pulsar folks have done on occasion.  Right, @savetheclocktower?

> Or simply change the grammar so that single statements are valid at the top level?

I think the Java grammar did this at one point for at least one construct.  I haven't had luck turning up where I read / saw that yet (^^;

**Comment by ObserverOfTime** [comment]
> I think the Java grammar did this at one point for at least one construct. I haven't had luck turning up where I read / saw that yet (^^;

You're probably thinking of [this](https://github.com/github/semantic/blob/main/docs/grammar-development-guide.md#know-when-to-parse-invalid-syntax).

**Comment by sogaiu** [comment]
Oh yes, that is likely it.  Thank you!

The relevant text seems to have been added [a fair bit back](https://github.com/github/semantic/commit/c7a1f1d30720f8709b2f4f69d02edab0288ea29a):

> While the goal is to emulate valid syntax, certain use-cases require supporting invalid syntax. There is a difference between parsing verses interpreting or compiling a language. Your parse tree doesn't always represent code that will run, but it has to be "valid enough" to represent readable code.

> As an example, the Java language requires statements (such as the expression `x = 3`) to be inside of methods. However, in the `tree-sitter-java` grammar, we explicitly allow statements to be parsed outside of methods. The grammar was loosened so that syntax highlighting for a code snippet of Java in a markdown file may be supported. While this code wouldn't compile, it doesn't break for the use-cases that rely on our parsing of the language. 

No wonder I wasn't finding it...completely different repository (^^;

**Comment by savetheclocktower** [comment]
> I think this kind of thing is something the Pulsar folks have done on occasion. Right, @savetheclocktower?

Almost. We've written very small Tree-sitter grammars for highlighting TODOs and hyperlinks in comments. We haven't ever written a grammar that would treat a “contextually valid” snippet (i.e., “if it were within [some context], it _would_ be valid”) as valid.

I'd be interested in doing that if the pattern weren't painful. I took 20 minutes to play around with a grammar that implemented Example 1 — formally extending the JavaScript grammar, but with `formal_parameters` as the root node — and I wasn't able to get very far. I had a hard time envisioning what would need to change in the parser so that it didn't consider anything that wasn't relevant to that limited role; it parsed `{ foo, bar }` exactly like its parent grammar would've.

> Oh yes, that is likely it. Thank you!

Good to know there's prior art there. I've long felt that `tree-sitter-css` should be doing this; folks share fragments of CSS in blog posts and Markdown documents all the time.

---

I do agree with @clason’s comment [elsewhere](https://github.com/tree-sitter-grammars/tree-sitter-svelte/issues/10#issuecomment-2348135538):

> I said it before, I'll say it again: Tree-sitter is simply a poor fit for templating languages, and you'll be fighting the tool all the way.

One illustration of that truth is that the examples listed in the description have a lot in common, but would have vastly different solutions if we didn't get the feature we're proposing here.

For instance:

* Example 3 (CSS) would benefit from a laxer parser (for the reasons explained above.
* Example 1a (TypeScript/Svelte) probably wouldn't! I doubt `tree-sitter-typescript` would be interested in treating a top-level `T extends string` as valid just because it solves Svelte's use case.
  
  In that instance, the syntax is simple enough that I bet we could get away with writing our own tiny single-purpose grammar just to parse it. It would be based on `tree-sitter-typescript` but it probably wouldn't have to change very often. (I hear the distant cackling of the gods as I type those words.)

* Example 1b (`#snippet`) is the toughest because the syntax in question is _valid_ in JavaScript on its own; it just has a different meaning when removed from its context. `{ foo, bar }` is parsed as two inert variable references inside a statement block; `({ foo, bar })` is interpreted as an object literal definition with shorthand property syntax; `function foo({ foo, bar })` is interpreted as a function definition with two destructured object properties being assigned to variables of identical names.

    One solution to the `#snippet` problem would involve outsourcing the parsing to `tree-sitter-javascript`, but overriding (at least some of) its highlighting logic with custom query statements. This is basically item 3 in [the description of this ticket](https://github.com/tree-sitter-grammars/tree-sitter-svelte/issues/10). That turns it into an editor concern; the editor would need to allow someone to share query files across packages. Amazingly, that's the _least_ hacky solution I can think of right now.

**Comment by savetheclocktower** [comment]
Well, hot damn — `tree-sitter-css` already treats bare declarations as valid:

<img width="681" alt="Screenshot 2024-09-13 at 10 47 36 AM" src="https://github.com/user-attachments/assets/52a5159b-de61-4b9c-9257-050b0f45246d">

In retrospect, I should've realized this; it's injected into CSS-in-JS contexts [like `styled-components`](https://styled-components.com/docs/basics#getting-started), and it wouldn't be any good at that job otherwise.

**Comment by maxbrunsfeld** [comment]
I do think that down the road, there should be an ability to parse a select node type from a grammar, instead of the root node. I believe it would require some additional generation-time work for *each* possible sub-rule that you wanted to allow this for. For example, the typescript grammar would have to explicitly allow type parameters to be used as a start rule, in order to support svelte.

In other words, I think that your Second solution should be on our roadmap (no promises on timelines though). I’m not in favor of your first solution, because I think all of these cases are better handled via injections, as opposed to trying to parse multiple languages in a single pass.

thanks @AlbertMarashi  for writing this up, and everyone else for the discussion.

**Comment by AlbertMarashi** [comment]
@maxbrunsfeld thanks. The second solution would likely work in most cases, my only gripe and concern with it would be how it relates to version control.

For example, if zed updates it's typescript treesitter which in turn breaks the node that the svelte extension was targeting within it's `highlights.scm` would that create issues?

**Comment by AlbertMarashi** [comment]
> I doubt tree-sitter-typescript would be interested in treating a top-level T extends string as valid just because it solves Svelte's use case

Agree @savetheclocktower, nor would I even want them to really add that.

**Comment by clason** [comment]
Yes, but that's a general problem not limited to that proposal. _Every_ grammar update can break queries, which must be adapted. This just adds one more point of failure to an already existing, large, set.

**Comment by AlbertMarashi** [comment]
> Yes, but that's a general problem not limited to that proposal. _Every_ grammar update can break queries, which must be adapted. This just adds one more point of failure to an already existing, large, set.

I suppose, although in solution 1 there would be a more explicit linking between languages and versions

**Comment by AlbertMarashi** [comment]
@ObserverOfTime btw your github bio cracked me up

**Comment by savetheclocktower** [comment]
> For example, if zed updates it's typescript treesitter which in turn breaks the node that the svelte extension was targeting within it's `highlights.scm` would that create issues?

It would, but this is the trade-off. Solution 1 is more self-contained and comprehensive; Solution 2 is more modular and streamlined. Each one has its drawbacks, so the better solution is the one that harmonizes most with the design and “core values” (ugh) of the project.

One way to mitigate the drawbacks of Solution 2 is to make versioning explicit and able to be inspected at runtime. If someone writes a grammar that might coexist between two different versions of `tree-sitter-typescript`, it could use the version number to decide how to configure the injection. (Probably by having two different `injections.scm`s to choose between.)

**Comment by clason** [comment]
> One way to mitigate the drawbacks of Solution 2 is to make versioning explicit and able to be inspected at runtime.

That would mitigate _a lot_ of drawbacks; it's something I've been requesting for a long time... 

(Of course, that would require parses to actually make versioned releases, which is still the exception. But it's a chicken-and-egg problem: As long as the version is basically useless, grammars will not be motivated to do that. So this is all just another instance of the same fundamental issue that's been around and known.)

**Comment by sogaiu** [comment]
> So this is all just another instance of the same fundamental issue that's been around and known.

Yes, though since earlier discussions like [this one](https://github.com/tree-sitter/tree-sitter/issues/2611), time has passed and use of tree-sitter has spread more.  It doesn't seem unreasonable to guess that the matter is affecting more folks and will continue to do so.

I think the recent discussion on discord / matrix with @casouri (Emacs tree-sitter person) are further evidence of this.

---

On a somewhat related note (as it relates to selecting an appropriate grammar), there is also the (not as big perhaps) matter of different grammar implementations for a particular language using the same `name`.  IIUC, this complicates os distribution packaging, but also makes use of multiple imlpementations within a single program (e.g. an editor) problematic.

The clashing has cropped up a few times [1] now and it seems likely that the number of incidents will increase over time, though not that quickly perhaps.

---

[1] [Here](https://github.com/casouri/tree-sitter-module/pull/50) is one example.

**Comment by AlbertMarashi** [comment]
I personally am quite a fan of solution 1, because the versioning contract is more explicit, and the code would be much easier to maintain.

However, that being said there is still a need for arbitrary language injections (like markdown code blocks) and I don't think that should go away.

In a nutshell, the benefits of allowing treesitter grammars to target arbitrary nodes in other languages is huge, and the use-cases would be quite common.

We all know how often as programmers we need to write doc-comments, markdown code blocks that are targeting a specific node type of a language that might not be a top-level rule.
